src/main/tetris_board.cpp

```diff
@@ -190,3 +247,115 @@ bool TetrisBoard::check_corners( TetrisPiece& piece, int direction_unit )
   }
   return can_shift;
 }
+
+vector<int> TetrisBoard::determine_full_rows()
+{
+  vector<int> full_rows;
+  for( size_t row = 0; row < occupied_spaces.size(); row++ )
+  {
+    bool full_row = true;
+    for( bool square_occupied : occupied_spaces.at( row ) )
+    {
+      if( !square_occupied )
+      {
+        full_row = false;
+      }
+    }
+    
+    if( full_row )
+    {
+      full_rows.push_back( row );
+    }
+  }
+  return full_rows;
+}
+
+void TetrisBoard::eliminate_full_rows( vector<int> full_rows )
+{
+  vector<unique_ptr<RenderComponent>> new_components;
+  for( auto& component : components )
+  {
+    try
+    {
+      Block& block = dynamic_cast<Block&>( *component );
+      vector<int>::iterator iterator = std::find( 
+        full_rows.begin(), 
+        full_rows.end(), 
+        block.get_row() );
+
+      if( iterator == full_rows.end() )
+      {
+        new_components.push_back( move( component ) );
+      }
+      else
+      {
+        occupied_spaces.at( block.get_row() ).at( block.get_column() ) = false;
+      }
+    }
+    catch( std::bad_cast )
+    {
+    }
+  }
+  components = move( new_components );
+}
+
+void TetrisBoard::update_blocks( std::vector<int> full_rows )
+{
+  if( !full_rows.empty() )
+  {
+    sort_blocks_by_row();  
+  }
+  
+  for( auto& component : components )
+  {
+    try
+    {
+      Block& block = dynamic_cast<Block&>( *component );
+      for( int row : full_rows )
+      {
+        if( block.get_row() < row )
+        {
+          while( !has_landed( block ) )
+          {
+            occupied_spaces.at( block.get_row() ).at( block.get_column() ) = false;
+            block.fall();  
```

Comment:
Here the block streaks though space, maintaining correct occupied space information as it falls. Not entirely sure if tetris blocks fall until they hit something, or if they only move the number of rows that were eliminated.